File: wakakusa/demo/links.rb
# Objects that point at one another. Yokan needs a weak back pointer
# here so a parent and a child cannot own each other forever; Ruby's
# collector takes a cycle in its stride, so the back pointer is an
# ordinary reference and the tree is written the way it reads.
require "wakakusa"

class Node
  attr_accessor :label, :kid, :parent

  def initialize(label)
    @label = label
    @kid = nil
    @parent = nil
  end
end

class Tree
  def initialize
    @root = nil
    @keep = nil
    @note = "-"
  end

  def build
    a = Node.new("alpha")
    b = Node.new("beta")
    a.kid = b
    b.parent = a
    @root = a
    @keep = b
  end

  def peek
    @note = if @root.nil?
              @keep.nil? ? "no root" : "kept #{@keep.label}, parent=#{@keep.parent.label}"
            elsif @root.kid.nil?
              "no kid"
            else
              "kid=#{@root.kid.label} parent=#{@root.kid.parent.label}"
            end
  end

  def view
    column(spacing: 8.0, padding: 12.0) {
      text "note: #{@note}"
      if @root.nil?
        text "root: (none)"
      else
        text "root: #{@root.label}"
      end
      row(spacing: 6.0) {
        button("build") { build }
        button("peek") { peek }
        button("drop") { @root = nil }
      }
    }
  end
end

run(Tree.new, title: "links")

File: gomamochi/demo/links.go
// Objects that point at one another. Yokan needs a weak back pointer
// here so a parent and a child cannot own each other forever; Go's
// collector takes a cycle in its stride, so the back pointer is an
// ordinary pointer and the tree is written the way it reads.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Node struct {
	label  string
	kid    *Node
	parent *Node
}

type Tree struct {
	root *Node
	keep *Node
	note string
}

func (t *Tree) build() {
	a := &Node{label: "alpha"}
	b := &Node{label: "beta"}
	a.kid = b
	b.parent = a
	t.root = a
	t.keep = b
}

func (t *Tree) peek() {
	switch {
	case t.root == nil && t.keep == nil:
		t.note = "no root"
	case t.root == nil:
		t.note = fmt.Sprintf("kept %s, parent=%s", t.keep.label, t.keep.parent.label)
	case t.root.kid == nil:
		t.note = "no kid"
	default:
		t.note = fmt.Sprintf("kid=%s parent=%s", t.root.kid.label, t.root.kid.parent.label)
	}
}

func (t *Tree) View() Element {
	rootLine := Text("root: (none)")
	if t.root != nil {
		rootLine = Text("root: " + t.root.label)
	}
	return Column(
		Text("note: "+t.note),
		rootLine,
		Row(
			Button("build").OnClick(func() { t.build() }),
			Button("peek").OnClick(func() { t.peek() }),
			Button("drop").OnClick(func() { t.root = nil }),
		).Spacing(6),
	).Spacing(8).Padding(12)
}

func main() {
	Run(&Tree{note: "-"}, Title("links"))
}
